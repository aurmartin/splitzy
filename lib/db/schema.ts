import { type RunResult } from "better-sqlite3";
import { type ExtractTablesWithRelations } from "drizzle-orm";
import { type BetterSQLite3Database } from "drizzle-orm/better-sqlite3";
import { type ExpoSQLiteDatabase } from "drizzle-orm/expo-sqlite";
import {
  SQLiteColumnBuilderBase,
  SQLiteTransaction,
  sqliteTable,
  text,
} from "drizzle-orm/sqlite-core";
import { type SQLiteRunResult } from "expo-sqlite";

export const groupsTable = sqliteTable("groups", {
  id: text().primaryKey(),
  syncStatus: text().notNull(),
  name: text().notNull(),
  currency: text().notNull(),
  members: text().notNull(),
  createdAt: text().notNull(),
  updatedAt: text().notNull(),
  deletedAt: text(),
});

export const expensesTable = sqliteTable("expenses", {
  id: text().primaryKey(),
  syncStatus: text().notNull(),
  groupId: text().notNull(),
  title: text().notNull(),
  payerName: text().notNull(),
  splitExpense: text().notNull(),
  createdAt: text().notNull(),
  updatedAt: text().notNull(),
  deletedAt: text(),
  receipt: text(),
});

export const syncedTable = (
  name: string,
  columns: Record<string, SQLiteColumnBuilderBase>,
) => {
  return sqliteTable(name, {
    ...columns,
    id: text().primaryKey(),
    syncStatus: text().notNull(),
  });
};

export type SyncableTable = ReturnType<typeof syncedTable>;

export const syncQueue = sqliteTable("sync_queue", {
  id: text().primaryKey(),
  type: text().notNull(),
  table: text().notNull(),
  objectId: text().notNull(),
  data: text(),
  createdAt: text().notNull(),
});

export const tables = {
  groups: groupsTable,
  expenses: expensesTable,
};

export type ExpenseRecord = typeof expensesTable.$inferSelect;
export type GroupRecord = typeof groupsTable.$inferSelect;
export type SyncQueueRecord = typeof syncQueue.$inferSelect;

export type Transaction = SQLiteTransaction<
  "sync",
  SQLiteRunResult | RunResult,
  Record<string, unknown>,
  ExtractTablesWithRelations<Record<string, unknown>>
>;

export type Database = ExpoSQLiteDatabase | BetterSQLite3Database;
