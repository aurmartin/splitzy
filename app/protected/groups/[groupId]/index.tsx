import Button from "@/components/button";
import Card from "@/components/card";
import FAB from "@/components/fab";
import { Pressable } from "@/components/pressable";
import { Screen } from "@/components/screen";
import { useSnackBar } from "@/components/snack-bar";
import { useSystem } from "@/components/system-provider";
import { Text } from "@/components/text";
import { TopBar } from "@/components/top-bar";
import { useBalance } from "@/lib/balance";
import {
  getAbsoluteAmount,
  getAmountColor,
  getBalanceMessage,
} from "@/lib/currency";
import { tables } from "@/lib/db/schema";
import { Env } from "@/lib/env";
import { Expense, useExpenses } from "@/lib/expenses";
import { Group, useGroup, useMe, getMe } from "@/lib/groups";
import { getLocale } from "@/lib/locale";
import { Ionicons } from "@expo/vector-icons";
import { Redirect, router, useLocalSearchParams, useRouter } from "expo-router";
import React from "react";
import { SectionList, Share, View } from "react-native";

interface ExpenseSection {
  title: string;
  data: Expense[];
}

const ExpenseItem = (props: { group: Group; expense: Expense }) => {
  const { group, expense } = props;
  const router = useRouter();

  return (
    <Card
      key={expense.id}
      onPress={() =>
        router.navigate(`/protected/groups/${group.id}/expenses/${expense.id}`)
      }
    >
      <View
        style={{
          flexDirection: "row",
          justifyContent: "space-between",
          alignItems: "center",
        }}
      >
        <View>
          <Text type="title" style={{ marginBottom: 8 }}>
            {expense.title}
          </Text>
          <Text style={{ color: "hsl(0 0% 40%)" }}>
            {expense.payerName}, {expense.createdAt.toLocaleDateString()}
          </Text>
        </View>

        <Text>
          {expense.splitExpense.total.setLocale(getLocale()).toFormat()}
        </Text>
      </View>
    </Card>
  );
};

const Expenses = React.memo(function Expenses(props: { group: Group }) {
  const { group } = props;
  const expenses = useExpenses(group.id);
  const system = useSystem();
  const [refreshing, setRefreshing] = React.useState(false);
  const snackBar = useSnackBar();

  const formatDate = (date: Date) => {
    const today = new Date();
    const yesterday = new Date(today);
    yesterday.setDate(yesterday.getDate() - 1);

    if (date.toDateString() === today.toDateString()) {
      return "Aujourd'hui";
    } else if (date.toDateString() === yesterday.toDateString()) {
      return "Hier";
    } else {
      return date.toLocaleDateString(undefined, {
        weekday: "long",
        year: "numeric",
        month: "long",
        day: "numeric",
      });
    }
  };

  const sections = React.useMemo(() => {
    const sections: Record<string, Expense[]> = {};

    expenses.forEach((expense) => {
      const date = new Date(expense.createdAt);
      const formattedDate = formatDate(date);

      if (!sections[formattedDate]) {
        sections[formattedDate] = [];
      }
      sections[formattedDate].push(expense);
    });

    return Object.entries(sections).map(([date, ids]) => ({
      title: date,
      data: ids,
    }));
  }, [expenses]);

  const renderItem = ({ item }: { item: Expense }) => (
    <ExpenseItem group={group} expense={item} />
  );

  const renderSectionHeader = ({ section }: { section: ExpenseSection }) => (
    <Text style={{ fontWeight: "bold" }}>{section.title}</Text>
  );

  const onRefresh = React.useCallback(async () => {
    setRefreshing(true);

    try {
      await system.syncEngine.syncTableFromRemote(tables.expenses);
    } catch {
      snackBar.show(
        "Une erreur est survenue lors de la synchronisation des dépenses.",
        "error",
      );
    } finally {
      setRefreshing(false);
    }
  }, [system, snackBar]);

  return (
    <SectionList
      testID="expenses-list"
      contentContainerStyle={{ gap: 8 }}
      sections={sections}
      renderItem={renderItem}
      renderSectionHeader={renderSectionHeader}
      maxToRenderPerBatch={10}
      initialNumToRender={10}
      windowSize={10}
      onRefresh={onRefresh}
      refreshing={refreshing}
      style={{ borderRadius: 8 }}
      ListFooterComponent={<View style={{ height: 80 }} />}
      ListEmptyComponent={<Text>Aucune dépense</Text>}
      showsVerticalScrollIndicator={false}
    />
  );
});

const BalanceDisplay = React.memo(function BalanceDisplay(props: {
  group: Group;
}) {
  const { group } = props;
  const balance = useBalance(group);
  const me = useMe(group.id);

  const myBalance = balance[me];

  return (
    <View
      style={{
        alignItems: "center",
        backgroundColor: "white",
        padding: 16,
        borderRadius: 10,
        marginBottom: 16,
      }}
    >
      <Text type="title">{getBalanceMessage(myBalance)}</Text>
      {myBalance.isZero() ? null : (
        <Text type="title" style={{ color: getAmountColor(myBalance) }}>
          {getAbsoluteAmount(myBalance).setLocale(getLocale()).toFormat()}
        </Text>
      )}
      <Button
        onPress={() => router.navigate(`/protected/groups/${group.id}/balance`)}
        style={{ marginTop: 8 }}
      >
        Remboursements
      </Button>
    </View>
  );
});

const GroupScreen = React.memo(function GroupScreen() {
  const router = useRouter();
  const { groupId }: { groupId: string } = useLocalSearchParams();
  const group = useGroup(groupId);
  const system = useSystem();

  const shareGroup = React.useCallback(() => {
    if (!group) {
      return;
    }

    const getLink = () => {
      switch (Env.APP_ENV) {
        case "development":
          return `https://splitzy.aurmartin.fr/dev/groups/${group.id}/import`;
        case "preview":
          return `https://splitzy.aurmartin.fr/preview/groups/${group.id}/import`;
        case "production":
          return `https://splitzy.aurmartin.fr/groups/${group.id}/import`;
      }
    };

    const link = getLink();

    Share.share({
      title: `Rejoins le groupe "${group.name}" sur Splitzy !`,
      message: `Hey ! Je t'invite à rejoindre mon groupe "${group.name}" sur Splitzy pour gérer nos dépenses partagées. Clique sur ce lien pour nous rejoindre !\n\n${link}`,
    });
  }, [group]);

  if (!group) {
    throw new Error("Group not found");
  }

  if (getMe(system, groupId) === null) {
    return <Redirect href={`/protected/groups/${groupId}/set-me`} />;
  }

  return (
    <Screen style={{ paddingBottom: 0 }}>
      <TopBar
        title={group.name}
        rightActions={[
          <Pressable
            testID="share-group"
            key="share"
            style={{
              backgroundColor: "white",
              padding: 8,
              borderRadius: 24,
            }}
            onPress={shareGroup}
            android_ripple={{ color: "hsl(0 0% 90%)" }}
          >
            <Ionicons name="share-outline" size={24} color="hsl(0 0% 40%)" />
          </Pressable>,
        ]}
      />

      <BalanceDisplay group={group} />

      <Expenses group={group} />

      <FAB
        onPress={() =>
          router.navigate(`/protected/groups/${group.id}/new-expense`)
        }
      >
        <Ionicons name="add-outline" size={24} color="white" />
      </FAB>
    </Screen>
  );
});

export default GroupScreen;
