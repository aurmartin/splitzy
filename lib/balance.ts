import dinero, { type Dinero } from "dinero.js";
import { type Expense, ExpenseService, useExpenses } from "@/lib/expenses";
import { type Group } from "@/lib/groups";

export type Balance = Record<string, Dinero>;

export type Reimbursement = {
  from: string;
  to: string;
  amount: Dinero;
};

export function computeBalance(group: Group, expenses: Expense[]): Balance {
  const balance = group.members.reduce((acc, member) => {
    acc[member] = dinero({ amount: 0, currency: group.currency });
    return acc;
  }, {} as Balance);

  expenses.forEach((expense) => {
    const payer = expense.payerName;
    const total = expense.splitExpense.total;

    balance[payer] = balance[payer].add(total);

    const amounts = ExpenseService.getAmounts(expense);

    Object.entries(amounts).forEach(([name, amount]) => {
      balance[name] = balance[name].subtract(amount);
    });
  });

  return balance;
}

export const useBalance = (group: Group): Balance => {
  const expenses = useExpenses(group.id);
  return computeBalance(group, expenses);
};

export function computeOptimalReimbursements(
  balance: Balance,
): Reimbursement[] {
  const reimbursements: Reimbursement[] = [];
  const members = Object.keys(balance);

  // Create a copy of the balance object
  const workingBalance = members.reduce((acc, member) => {
    acc[member] = balance[member];
    return acc;
  }, {} as Balance);

  // Continue while there are non-zero balances
  while (members.some((member) => !workingBalance[member].isZero())) {
    // Find the person who owes the most (most negative balance)
    const maxDebtor = members.reduce((a, b) =>
      workingBalance[a].lessThan(workingBalance[b]) ? a : b,
    );

    // Find the person who is owed the most (most positive balance)
    const maxCreditor = members.reduce((a, b) =>
      workingBalance[a].greaterThan(workingBalance[b]) ? a : b,
    );

    // If both balances are effectively zero (less than 1 cent), break
    if (
      workingBalance[maxDebtor].isZero() &&
      workingBalance[maxCreditor].isZero()
    ) {
      break;
    }

    // Calculate the reimbursement amount (minimum of absolute values)
    const debtorAmount = workingBalance[maxDebtor].multiply(-1); // Make positive
    const creditorAmount = workingBalance[maxCreditor];
    const amount = debtorAmount.lessThan(creditorAmount)
      ? debtorAmount
      : creditorAmount;

    // Update the balances
    workingBalance[maxDebtor] = workingBalance[maxDebtor].add(amount);
    workingBalance[maxCreditor] = workingBalance[maxCreditor].subtract(amount);

    // Record the reimbursement
    reimbursements.push({
      from: maxDebtor,
      to: maxCreditor,
      amount,
    });
  }

  return reimbursements;
}
