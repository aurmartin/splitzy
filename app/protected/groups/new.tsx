import Button from "@/components/button";
import Label from "@/components/label";
import TextInput from "@/components/text-input";
import { useAddGroup } from "@/lib/groups";
import { Ionicons } from "@expo/vector-icons";
import { Stack, useRouter } from "expo-router";
import React from "react";
import { View } from "react-native";
import { ValidationError } from "@/lib/validation-error";
import { Screen } from "@/components/screen";
import { TopBar } from "@/components/top-bar";
import { Text } from "@/components/text";

const DEFAULT_CURRENCY = "EUR";

type PromiseState = {
  loading: boolean;
  validationErrors?: Record<string, string>;
  error?: string;
};

type Action =
  | { type: "start" }
  | { type: "validationError"; payload: Record<string, string> }
  | { type: "error"; payload: string }
  | { type: "success" };

function promiseStateReducer(_state: PromiseState, action: Action) {
  switch (action.type) {
    case "start":
      return { loading: true };
    case "validationError":
      return { loading: false, validationErrors: action.payload };
    case "error":
      return { loading: false, error: action.payload };
    case "success":
      return { loading: false };
  }
}

const GroupForm = () => {
  const router = useRouter();
  const addGroup = useAddGroup();

  const [name, setName] = React.useState("");
  const [members, setMembers] = React.useState<string[]>([""]);

  const [promiseState, dispatch] = React.useReducer(promiseStateReducer, {
    loading: false,
  });

  const handleSubmit = async () => {
    try {
      dispatch({ type: "start" });

      await addGroup({ name, currency: DEFAULT_CURRENCY, members });

      dispatch({ type: "success" });

      router.navigate("/");
    } catch (error) {
      if (error instanceof ValidationError) {
        dispatch({ type: "validationError", payload: error.errors });
      } else {
        console.error("Error adding group", error);

        const errorMessage =
          error instanceof Error ? error.message : "Une erreur est survenue";

        dispatch({ type: "error", payload: errorMessage });
      }
    }
  };

  const addMember = () => {
    setMembers([...members, ""]);
  };

  const changeMember = (index: number, member: string) => {
    setMembers(members.map((m, i) => (i === index ? member : m)));
  };

  const removeMember = (index: number) => {
    if (index === 0) return;
    setMembers(members.filter((_, i) => i !== index));
  };

  return (
    <Screen>
      <TopBar title="Nouveau groupe" />

      <View style={{ flex: 1, gap: 16 }}>
        {promiseState.error && (
          <Text style={{ color: "hsl(0, 100%, 30%)" }}>
            {promiseState.error}
          </Text>
        )}

        <View>
          <TextInput
            value={name}
            onChangeText={setName}
            label="Nom du groupe"
            placeholder="Nom du groupe"
          />
          {promiseState.validationErrors?.name && (
            <Text style={{ color: "hsl(0, 100%, 30%)" }}>
              {promiseState.validationErrors.name}
            </Text>
          )}
        </View>

        <View>
          <Label>Membres</Label>
          <View
            style={{
              gap: 8,
              borderRadius: 4,
            }}
          >
            {members.map((member, index) => (
              <View
                key={index}
                style={{
                  flexDirection: "row",
                  alignItems: "center",
                  gap: 8,
                }}
              >
                <TextInput
                  style={{ flex: 1 }}
                  value={member}
                  onChangeText={(text) => changeMember(index, text)}
                  placeholder="Nouveau membre"
                />

                <Ionicons
                  name="trash-outline"
                  size={20}
                  color={index === 0 ? "hsl(0 0% 50%)" : "black"}
                  onPress={() => removeMember(index)}
                />
              </View>
            ))}
            <Button onPress={addMember} type="secondary">
              Ajouter un membre
            </Button>
          </View>
          {promiseState.validationErrors?.members && (
            <Text style={{ color: "hsl(0, 100%, 30%)" }}>
              {promiseState.validationErrors.members}
            </Text>
          )}
        </View>

        <Button onPress={handleSubmit} style={{ marginTop: 16 }}>
          Cr√©er le groupe
        </Button>
      </View>
    </Screen>
  );
};

export default function NewGroupScreen() {
  return (
    <View style={{ flex: 1 }}>
      <Stack.Screen options={{ title: "New Group" }} />
      <GroupForm />
    </View>
  );
}
