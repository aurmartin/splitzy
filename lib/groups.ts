import { useSystem } from "@/components/system-provider";
import { ValidationError } from "@/lib/validation-error";
import { GroupRecord, groupsTable } from "@/lib/db/schema";
import { System } from "@/lib/system";
import { generateId, time } from "@/lib/utils";
import { Currency } from "dinero.js";
import { desc, eq } from "drizzle-orm";
import { useLiveQuery } from "drizzle-orm/expo-sqlite";
import { useCallback } from "react";

export interface Group {
  id: string;
  name: string;
  currency: Currency;
  members: string[];

  createdAt: Date;
  updatedAt: Date;
  deletedAt: Date | undefined;
}

export interface GroupCreateParams {
  name: string;
  currency: Currency;
  members: string[];
}

export const useAddGroup = () => {
  const system = useSystem();

  return useCallback(
    async (params: GroupCreateParams) => addGroup(system, params),
    [system],
  );
};

const addGroup = (system: System, params: GroupCreateParams) =>
  time("addGroup", () => _addGroup(system, params));

const _addGroup = async (system: System, params: GroupCreateParams) => {
  params = checkGroupCreateParams(params);

  const id = generateId();

  try {
    const fields = {
      id,
      name: params.name,
      currency: params.currency,
      members: JSON.stringify(params.members),
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
      deletedAt: null,
    };

    await system.syncEngine.insert(groupsTable, id, fields);
  } catch (error) {
    console.error("Error creating group", error);
    throw error;
  }
};

export const useImportGroup = () => {
  const system = useSystem();

  return useCallback(
    async (groupId: string) => {
      const user = await system.supabaseConnector.client.auth.getUser();

      await system.supabaseConnector.client.from("members").insert({
        user_id: user.data.user!.id,
        group_id: groupId,
      });

      await system.syncEngine.syncTableFromRemote(groupsTable);
    },
    [system],
  );
};

const checkGroupCreateParams = (params: GroupCreateParams) => {
  const errors: Record<string, string> = {};

  params.members = params.members.filter((member) => member.length > 0);

  if (params.name.length === 0) {
    errors.name = "Veuillez entrer un nom";
  }

  if (params.members.length === 0) {
    errors.members = "Veuillez entrer au moins un membre";
  }

  if (Object.keys(errors).length > 0) {
    throw new ValidationError(errors);
  }

  return params;
};

const _useGroups = (): Group[] => {
  const system = useSystem();
  const { data } = useLiveQuery(
    system.db.select().from(groupsTable).orderBy(desc(groupsTable.createdAt)),
  );

  return data.map(decodeGroup);
};

export const useGroups = () => time("useGroups", _useGroups);

export const useGroup = (groupId: string): Group | undefined => {
  const system = useSystem();
  return getGroupRow(system, groupId);
};

export const useMe = (groupId: string): string => {
  const system = useSystem();
  const me = getMe(system, groupId);

  if (!me) {
    throw new Error("Me not found");
  }

  return me;
};

export const getMe = (system: System, groupId: string) =>
  system.kvStorage.getItem(`me:${groupId}`);

export const setMe = (system: System, groupId: string, me: string) =>
  system.kvStorage.setItem(`me:${groupId}`, me);

export const removeMe = (system: System, groupId: string) =>
  system.kvStorage.removeItem(`me:${groupId}`);

export const getGroupRow = (
  system: System,
  groupId: string,
): Group | undefined => {
  const group = system.db
    .select()
    .from(groupsTable)
    .where(eq(groupsTable.id, groupId))
    .limit(1)
    .all();

  if (group.length === 0) {
    return undefined;
  }

  return decodeGroup(group[0]);
};

const decodeGroup = (groupRecord: GroupRecord): Group => {
  return {
    id: groupRecord.id,
    name: groupRecord.name as string,
    members: JSON.parse(groupRecord.members as string),
    currency: groupRecord.currency as Currency,
    createdAt: new Date(groupRecord.createdAt as string),
    updatedAt: new Date(groupRecord.updatedAt as string),
    deletedAt: groupRecord.deletedAt
      ? new Date(groupRecord.deletedAt as string)
      : undefined,
  };
};
