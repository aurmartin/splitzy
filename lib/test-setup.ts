import { System } from "@/lib/system";
import {
  clearDatabase,
  createDatabase,
  createSupabaseServer,
} from "@/lib/test-utils";

const server = createSupabaseServer();
let system: System;

beforeEach(async () => {
  clearDatabase(system.db);
});

beforeAll(async () => {
  server.listen();
  system = new System(createDatabase());
  await system.initializeMigrations();
});

afterEach(() => {
  server.resetHandlers();
});

afterAll(() => {
  server.close();
  system.dispose();
});

export { server, system };
