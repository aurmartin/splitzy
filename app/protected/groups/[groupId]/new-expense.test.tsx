import NewExpenseScreen from "@/app/protected/groups/[groupId]/new-expense";
import { tables } from "@/lib/db/schema";
import { setMe } from "@/lib/groups";
import { system } from "@/lib/test-setup";
import { buildGroupRecord, renderRouter } from "@/lib/test-utils";
import { screen, userEvent, waitFor } from "@testing-library/react-native";
import { eq } from "drizzle-orm";

jest.mock("expo-document-picker", () => ({
  getDocumentAsync: jest.fn().mockResolvedValue({
    assets: [
      {
        uri: "https://example.com/receipt.jpg",
      },
    ],
  }),
}));

jest.mock("expo-file-system", () => ({
  ...jest.requireActual("expo-file-system"),
  readAsStringAsync: jest.fn().mockResolvedValue(""),
}));

beforeEach(async () => {
  const group = buildGroupRecord({ id: "test-group-id" });
  await system.db.insert(tables.groups).values(group);
  setMe(system, group.id, "Alice");
});

const routerContext = {
  "/protected/groups/[groupId]/new-expense": () => <NewExpenseScreen />,
};

const newExpenseScreenUrl = "/protected/groups/test-group-id/new-expense";

describe("NewExpenseScreen", () => {
  it("should render correctly", () => {
    renderRouter(routerContext, system, { initialUrl: newExpenseScreenUrl });
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("should create an expense", async () => {
    const user = userEvent.setup();

    renderRouter(routerContext, system, { initialUrl: newExpenseScreenUrl });

    await user.clear(screen.getByLabelText("Titre"));
    await user.type(screen.getByLabelText("Titre"), "Test expense");
    await user.type(screen.getAllByLabelText(/Montant/)[0], "100");
    await user.press(screen.getByRole("button", { name: /Alice/ }));
    await user.press(screen.getByLabelText("Enregistrer"));

    await waitFor(async () => {
      const expenses = await system.db
        .select()
        .from(tables.expenses)
        .where(eq(tables.expenses.groupId, "test-group-id"));

      expect(expenses).toHaveLength(1);
      expect(expenses[0].title).toBe("Test expense");
      expect(expenses[0].payerName).toBe("Alice");
    });
  });

  it("should show validation errors if the form is invalid", async () => {
    const user = userEvent.setup();

    renderRouter(routerContext, system, { initialUrl: newExpenseScreenUrl });

    await user.clear(screen.getByLabelText("Titre"));
    await user.press(screen.getByLabelText("Enregistrer"));

    await screen.findByText("Veuillez entrer un titre");
  });

  it("should throw an error if the group is not found", () => {
    expect(() =>
      renderRouter(routerContext, system, {
        initialUrl: "/protected/groups/not-found/new-expense",
      }),
    ).toThrow("Group not found");
  });

  it("should handle receipt", async () => {
    const user = userEvent.setup();

    renderRouter(routerContext, system, { initialUrl: newExpenseScreenUrl });

    await user.press(
      screen.getByRole("button", { name: /Télécharger un reçu/ }),
    );

    await screen.findByDisplayValue("Test receipt");
  });
});
