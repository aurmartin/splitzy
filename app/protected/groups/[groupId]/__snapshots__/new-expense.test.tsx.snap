// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewExpenseScreen should render correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNCSafeAreaView
    edges={
      {
        "bottom": "additive",
        "left": "additive",
        "right": "additive",
        "top": "additive",
      }
    }
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNCSafeAreaView
      edges={
        {
          "bottom": "additive",
          "left": "additive",
          "right": "additive",
          "top": "additive",
        }
      }
      style={
        [
          {
            "backgroundColor": "hsl(200 15% 90%)",
            "flex": 1,
            "padding": 8,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "marginBottom": 16,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "white",
              "borderRadius": 24,
              "padding": 8,
            }
          }
          testID="topbar-back-button"
        >
          <Text />
        </View>
        <Text
          style={
            [
              {
                "fontSize": 32,
                "fontWeight": "400",
                "lineHeight": 38,
              },
              {
                "marginLeft": 16,
              },
            ]
          }
        >
          Nouvelle dépense
        </Text>
      </View>
      <RCTScrollView>
        <View>
          <View
            style={
              {
                "flex": 1,
                "gap": 16,
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  [
                    {
                      "alignItems": "center",
                      "borderRadius": 40,
                      "height": 40,
                      "justifyContent": "center",
                      "paddingHorizontal": 16,
                    },
                    {
                      "backgroundColor": "hsl(200 40% 50%)",
                    },
                    undefined,
                  ],
                  {
                    "overflow": "hidden",
                  },
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "fontSize": 16,
                      "fontWeight": "400",
                      "lineHeight": 24,
                    },
                    {
                      "color": "white",
                    },
                  ]
                }
              >
                Télécharger un reçu
              </Text>
            </View>
            <View>
              <Text
                style={
                  [
                    {
                      "fontSize": 16,
                      "fontWeight": "400",
                      "lineHeight": 24,
                    },
                    {
                      "marginBottom": 4,
                    },
                  ]
                }
              >
                Titre
              </Text>
              <TextInput
                label="Titre"
                onChangeText={[Function]}
                placeholder="Titre"
                placeholderTextColor="#666"
                style={
                  [
                    {
                      "backgroundColor": "#fff",
                      "borderColor": "hsl(0 0% 70%)",
                      "borderRadius": 4,
                      "borderWidth": 1,
                      "height": 45,
                      "padding": 10,
                    },
                    undefined,
                  ]
                }
                value="Dépense"
              />
            </View>
            <View>
              <Text
                style={
                  [
                    {
                      "fontSize": 16,
                      "fontWeight": "400",
                      "lineHeight": 24,
                    },
                    {
                      "marginBottom": 4,
                    },
                  ]
                }
              >
                Montant (€)
              </Text>
              <TextInput
                keyboardType="numeric"
                onChangeText={[Function]}
                placeholder="0.00"
                placeholderTextColor="#666"
                style={
                  [
                    {
                      "backgroundColor": "#fff",
                      "borderColor": "hsl(0 0% 70%)",
                      "borderRadius": 4,
                      "borderWidth": 1,
                      "height": 45,
                      "padding": 10,
                    },
                    undefined,
                  ]
                }
                testID="float-input"
                value="0"
              />
            </View>
            <View>
              <View>
                <Text
                  style={
                    [
                      {
                        "fontSize": 16,
                        "fontWeight": "400",
                        "lineHeight": 24,
                      },
                      {
                        "marginBottom": 4,
                      },
                    ]
                  }
                >
                  Payé par
                </Text>
                <View
                  style={
                    {
                      "backgroundColor": "white",
                      "borderColor": "hsl(0 0% 70%)",
                      "borderRadius": 4,
                      "borderStyle": "solid",
                      "borderWidth": 1,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "marginBottom": -5,
                        "marginTop": -5,
                        "padding": 0,
                      }
                    }
                  >
                    <RNCPicker
                      items={
                        [
                          {
                            "label": "Alice",
                            "testID": undefined,
                            "textColor": undefined,
                            "value": "Alice",
                          },
                          {
                            "label": "Bob",
                            "testID": undefined,
                            "textColor": undefined,
                            "value": "Bob",
                          },
                        ]
                      }
                      numberOfLines={1}
                      onChange={[Function]}
                      selectedIndex={0}
                      style={
                        [
                          {
                            "height": 216,
                          },
                          undefined,
                        ]
                      }
                      testID="picker"
                    />
                  </View>
                </View>
              </View>
            </View>
            <View>
              <View
                style={
                  {
                    "gap": 8,
                  }
                }
              >
                <View>
                  <Text
                    style={
                      [
                        {
                          "fontSize": 16,
                          "fontWeight": "400",
                          "lineHeight": 24,
                        },
                        {
                          "marginBottom": 4,
                        },
                      ]
                    }
                  >
                    Partager la dépense
                  </Text>
                  <View
                    style={
                      {
                        "backgroundColor": "white",
                        "borderColor": "hsl(0 0% 70%)",
                        "borderRadius": 4,
                        "borderStyle": "solid",
                        "borderWidth": 1,
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "marginBottom": -5,
                          "marginTop": -5,
                          "padding": 0,
                        }
                      }
                    >
                      <RNCPicker
                        items={
                          [
                            {
                              "label": "Également",
                              "testID": undefined,
                              "textColor": undefined,
                              "value": "equal",
                            },
                            {
                              "label": "En pourcentage",
                              "testID": undefined,
                              "textColor": undefined,
                              "value": "percentage",
                            },
                            {
                              "label": "En montant",
                              "testID": undefined,
                              "textColor": undefined,
                              "value": "amount",
                            },
                          ]
                        }
                        numberOfLines={1}
                        onChange={[Function]}
                        selectedIndex={0}
                        style={
                          [
                            {
                              "height": 216,
                            },
                            undefined,
                          ]
                        }
                        testID="picker"
                      />
                    </View>
                  </View>
                </View>
                <View
                  style={
                    {
                      "flex": 1,
                      "gap": 8,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "display": "flex",
                        "flexDirection": "row",
                        "justifyContent": "space-between",
                      }
                    }
                  >
                    <Text
                      numberOfLines={1}
                      style={
                        [
                          {
                            "fontSize": 16,
                            "fontWeight": "400",
                            "lineHeight": 24,
                          },
                        ]
                      }
                    >
                      Alice
                    </Text>
                    <Text
                      style={
                        [
                          {
                            "fontSize": 16,
                            "fontWeight": "400",
                            "lineHeight": 24,
                          },
                        ]
                      }
                    >
                      $0.00
                    </Text>
                  </View>
                  <View
                    style={
                      {
                        "display": "flex",
                        "flexDirection": "row",
                        "justifyContent": "space-between",
                      }
                    }
                  >
                    <Text
                      numberOfLines={1}
                      style={
                        [
                          {
                            "fontSize": 16,
                            "fontWeight": "400",
                            "lineHeight": 24,
                          },
                        ]
                      }
                    >
                      Bob
                    </Text>
                    <Text
                      style={
                        [
                          {
                            "fontSize": 16,
                            "fontWeight": "400",
                            "lineHeight": 24,
                          },
                        ]
                      }
                    >
                      $0.00
                    </Text>
                  </View>
                </View>
              </View>
            </View>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  [
                    {
                      "alignItems": "center",
                      "borderRadius": 40,
                      "height": 40,
                      "justifyContent": "center",
                      "paddingHorizontal": 16,
                    },
                    {
                      "backgroundColor": "hsl(200 40% 50%)",
                    },
                    undefined,
                  ],
                  {
                    "overflow": "hidden",
                  },
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "fontSize": 16,
                      "fontWeight": "400",
                      "lineHeight": 24,
                    },
                    {
                      "color": "white",
                    },
                  ]
                }
              >
                Créer la dépense
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </RNCSafeAreaView>
  </RNCSafeAreaView>
</RNCSafeAreaProvider>
`;
